File: src/main/scala/com/ons/sml/businessMethods/methods/StandardError.scala
package com.ons.sml.businessMethods.methods

import com.ons.sml.businessMethods.impl.StandardErrorImpl
import org.apache.spark.sql.DataFrame

/** This is the class that is called by the object at the bottom.
  *
  * @param dfIn DataFrame - The dataframe that the standard error method will be applied to.
  */
class StandardError (val dfIn: DataFrame){

  if (dfIn == null ) throw new Exception("DataFrame cannot be null")

  import StandardErrorImpl._

  val defaultCol ="StandardError"

  /** THis function checks that all the arguments provided have a values in them instead of null, if there are
    * any nulls then this throws an error.
    *
    * @param arg1 String - This is the first argument that's passed into the standard error function.
    * @param arg2 String - This is the second argument that's passed into the standard error function.
    * @param arg3 String - This is the third argument that's passed into the standard error function.
    */
  private def mandatoryArgCheck(arg1:String, arg2:String, arg3:String) : Unit = {

    if ((arg1 == null) ||(arg2 == null) ||(arg3 == null)) throw new Exception("Missing mandatory argument")
  }

  /** This function first checks that the arguments given are all there, if any of them are null/missing then it
    * throws an exception. Then if decides which dataframe to use, the class one or the function one, if the function
    * one is there then it will take that one, otherwise it takes the class one.
    *
    * Next the newColName is checked to see if it is null, if it is then the defaultCol is used instead. You can also
    * leave this variable blank and it will choose the defaultCol.
    *
    * The dataframe is then checked to see if all the columns names provided are in the dataframe, if not then this
    * throws an error. If they are in the dataframe then the findStandardError is called to add the standard error
    * as a column in the dataframe.
    *
    * The standard error is calculated using the formula:
    *
    *                             {{{ stdError = sqrt{ (x/y)*(x-y) / (y-1) } * z  }}}
    *
    * @param df  DataFrame - The dataframe that the standard error method will be applied to.
    * @param xCol String - The column to be used as X in the calculation
    * @param yCol String - The column to be used as Y in the calculation
    * @param zCol String - The column to be used as Z in the calculation
    * @param newColName String - This is what the standard error column is called, it can be defaulted to StandardError
    * @return DataFrame
    */
  def stdErr1(df: DataFrame, xCol: String, yCol: String, zCol: String,newColName: String = defaultCol): DataFrame={
    mandatoryArgCheck(xCol, yCol, zCol)

    val dF = if (df == null) dfIn else df
    val newColName1 =  if (newColName==null) defaultCol else newColName
    dF.checkColNames(Seq(xCol, yCol, zCol)).findStandardError(newColName1, xCol, yCol, zCol)
  }
}

/** This object calls the class above, this is what is called in order to use the class.
  *
  */
object StandardError {

  def standardError(df: DataFrame) : StandardError = {new StandardError(df)}

}

File: python/sparkts/onsMethods/StandardError.py
from pyspark.sql import DataFrame


class StandardError():
    """

    This class is to calculate the standard error for input data based on formula:
    {{{ stdError = sqrt{ (x/y)*(x-y) / (y-1) } * z  }}} and x,y,z are columns in the input dataframe.

    It accepts the input dataframe, applies the above calculation and output the dataframe with a new column
    along calculated 'standardError' value.

    """
    defaultCol = 'StandardError'

    def __init__(self, df=None):
        """
        Instantiates a JavaStandardFactory from a DataFrame.

        :param df: Input dataframe.
        """

        if df is None:
            raise TypeError

        self._df = df

        self._jStdErr = self._df._sc._jvm.com.ons.sml.api.java.methods.JavaStandardErrorFactory.standardError(
            self._df._jdf)

    def __mandatoryArgumentCheck(self, arg1, arg2, arg3):
        """
        This function raise Type Error if any one of the input arg is None.

        :param arg1: Input argument 1
        :param arg2: Input argument 2
        :param arg3: Input argument 3
        """
        if (arg1 is None) | (arg2 is None) | (arg3 is None):
            raise TypeError('Missing mandatory argument')

    def stdErr1(self, df, xCol, yCol, zCol, outputCol=defaultCol):
        """

        This function will add an extra column on to a DataFrame containing the standard error
        This is calculated using the formula:
                {{{ stdError = sqrt{ (x/y)*(x-y) / (y-1) } * z  }}}

        :param df: Input dataframe
        :param xCol: The column to be used as X in the calculation
        :param yCol: The column to be used as Y in the calculation
        :param zCol: The column to be used as Z in the calculation
        :param outputCol: The name of the standard error column
        :return: Dataframe with a extra column for standardError
        """
        self.__mandatoryArgumentCheck(xCol, yCol, zCol)

        if df is None:
            df = self._df

        return DataFrame(self._jStdErr.stdErr1(df._jdf, xCol, yCol, zCol, outputCol), df.sql_ctx)

def standardError(df):
    """
    This function instantiates a StandardError by passing the input dataframe.

    :param df: The input dataframe.
    :return: Dataframe with a extra column for standardError
    """
    return StandardError(df)
